cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -602,43 +603,111 @@ func (m *McmManager) retry(fn func(ctx context.Context) (bool, error), resourceT
+	)
 
-	machineList, err := m.machineLister.Machines(m.namespace).List(labels.Everything())
+	selector = selector.Add(*req)
+	machineList, err := m.machineLister.Machines(m.namespace).List(selector)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Machine objects %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Machine objects %v for machinedeployment %q", err, machinedeployment.Name)
 	}
 
 	nodeList, err := m.nodeLister.List(labels.Everything())
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Nodes %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Nodes %v", err)
 	}
 
-	var nodes []string
+	var instances []cloudprovider.Instance
 	// Bearing O(n2) complexity, assuming we will not have lot of nodes/machines, open for optimisations.
 	for _, machine := range machineList {
-		if strings.Contains(machine.Name, md.Name) {
-			var found bool
-			for _, node := range nodeList {
-				if machine.Labels["node"] == node.Name {
-					nodes = append(nodes, node.Spec.ProviderID)
-					found = true
-					break
+		instance := cloudprovider.Instance{}
+		var found bool
+		for _, node := range nodeList {
+			if machine.Labels["node"] == node.Name {
+				// to ensure machine obj with VM but without node obj are marked as `Creating` cloudprovider.Instance
+				// NOTE: machine obj with a `notReady` node is not marked `Creating`
+				instance.Id = node.Spec.ProviderID
+				found = true
+				break
+			}
+		}
+		if !found {
+			// No k8s node found - either the VM has not registered yet or MCM is unable to fulfill the request.
+			// Report a special placeholder ID so that the autoscaler can track it as an unregistered node.
+			instance.Id = placeholderInstanceIDForMachineObj(machine.Name)
+			instance.Status = &cloudprovider.InstanceStatus{
+				State:     cloudprovider.InstanceCreating,
+				ErrorInfo: getErrorInfo(machine),
+			}
+		}
+		instances = append(instances, instance)
+	}
+	return instances, nil
+}
+
+func placeholderInstanceIDForMachineObj(name string) string {
+	return fmt.Sprintf("requested://%s", name)
+}
+
+// getErrorInfo returns cloudprovider.InstanceErrorInfo for the machine obj
+func getErrorInfo(machine *v1alpha1.Machine) *cloudprovider.InstanceErrorInfo {
+	if machine.Status.LastOperation.Type == v1alpha1.MachineOperationCreate && machine.Status.LastOperation.State == v1alpha1.MachineStateFailed && machine.Status.LastOperation.ErrorCode == machinecodes.ResourceExhausted.String() {
+		return &cloudprovider.InstanceErrorInfo{
+			ErrorClass:   cloudprovider.OutOfResourcesErrorClass,
+			ErrorCode:    machinecodes.ResourceExhausted.String(),
+			ErrorMessage: getMachineStatusErrorMessage(machine),
+		}
+	}
+	return nil
+}
+
+func getMachineStatusErrorMessage(machine *v1alpha1.Machine) string {
+	desc := machine.Status.LastOperation.Description
+	decoded, err := findCodeAndMessage(desc)
+	if err != nil {
+		return desc
```

Comment:
Can a case happen where `findCodeAndMessage` returns an error?  If so, shouldn't we return the error to the caller instead of returning the description?

Reply:
We are trying to get the message out of the description, which is a best-effort kind of thing.. if we are not able to get the message (lets say some message can't be parsed) then we should use the entire description... 
Returning error won't do much...